File: 38_sliding_windows/38_4_enduring_best_seller_streak.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

bool enduring_best_seller_streak(std::vector<std::string>& best_seller, int k) {
  int l = 0, r = 0;
  int cur = 0;
  std::unordered_map<std::string, int> hashmap;

  while (r < best_seller.size()) {
    if (hashmap.find(best_seller[r]) == hashmap.end()) {
      hashmap[best_seller[r]] = 0;
    }
    hashmap[best_seller[r]] += 1;
    cur += 1;

    if (cur == k) {
      if (hashmap[best_seller[r]] == k) {
        return true;
      }

      hashmap[best_seller[l]] -= 1;
      l += 1;
      cur -= 1;
    }

    r += 1;
  }
  return false;
}

int main() {
    std::vector<std::string> best_seller = {"book3", "book1", "book3", "book3", "book2"};
    int k = 3;
    std::cout << (enduring_best_seller_streak(best_seller, k) ? "true" : "false") << std::endl;

    best_seller = {"book3", "book1", "book3", "book3", "book2"};
    k = 2;
    std::cout << (enduring_best_seller_streak(best_seller, k) ? "true" : "false") << std::endl;

    return 0;
}

File: 38_sliding_windows/38_4_enduring_best_seller_streak.py
def enduring_best_seller_streak(best_seller: list, k: int)->bool:
  l, r = 0, 0
  cur = 0
  hashmap = {}
  while r < len(best_seller):
    if best_seller[r] not in hashmap:
      hashmap[best_seller[r]] = 0
    hashmap[best_seller[r]] += 1
    cur += 1
    if cur == k:
      if hashmap[best_seller[r]] == k:
        return True
      
      hashmap[best_seller[l]] -= 1
      l += 1
      cur -= 1
    
    r += 1

  return False

if __name__ == "__main__":
  best_seller = ["book3", "book1", "book3", "book3", "book2"]
  k = 3
  print(enduring_best_seller_streak(best_seller, k))
  best_seller = ["book3", "book1", "book3", "book3", "book2"]
  k = 2
  print(enduring_best_seller_streak(best_seller, k))